torch/csrc/autograd/functions/convolution.cpp

```diff
@@ -327,6 +365,126 @@ auto ConvBackward::releaseVariables() -> void {
   bias_.data.reset();
 }
 
+
+// ConvBackwardBackward implementation
+
+auto ConvBackwardBackward::apply(const variable_list& grad_grad_inputs) -> variable_list {
+  if (grad_grad_inputs.size() != 3) throw std::runtime_error("expected three grad_grad_inputs");
+  if (transposed) throw std::runtime_error("ConvBackwardBackward does not support transposed convolution");
+
+  auto ggI = grad_grad_inputs[0];
+  auto ggW = grad_grad_inputs[1];
+  auto ggb = grad_grad_inputs[2];
+
+  auto gO = grad_output_.unpack();
+  auto weight = weight_.unpack();
+
+  // Compute ggO = conv(w, ggI) + conv(ggW, i) + ggb
+  std::shared_ptr<Variable> ggO = nullptr;
+  if (ggI) {
+    ggO = std::make_shared<ConvForward>(*this)->apply({ggI, weight, nullptr})[0];
+  }
+
+  if (ggW) {
+    auto ggW_term = std::make_shared<ConvForward>(*this)->apply({input_.unpack(), ggW, nullptr})[0];
+    if (ggO) {
+      ggO = std::make_shared<Add>()->apply({ggO, ggW_term})[0];
+    } else {
+      ggO = ggW_term;
+    }
+  }
+
+  if (ggb) {
+    // View as (1, ggb.size(0), 1, 1...)
+    std::vector<long> new_size(gO->data->sizes().size(), 1);
+    new_size[1] = ggb->data->sizes()[0];
+    auto ggb_contiguous = std::make_shared<Clone>()->apply({ggb})[0];
+    auto ggb_view = std::make_shared<View>(new_size)->apply({ggb_contiguous})[0];
+
+    // Expand 
+    auto ggb_expanded = std::make_shared<Expand>(gO->data->sizes())->apply({ggb_view})[0];
+
+    if (ggO) {
+      ggO = std::make_shared<Add>()->apply({ggO, ggb_expanded})[0];
+    } else {
+      ggO = ggb_expanded;
+    }
+  }
+
+  // Compute gW = conv(gO, ggI)
+  std::shared_ptr<Variable> gW = nullptr;
+  if (ggI) {
+    // Modified params with correct padding
+    ConvParams gw_conv_params(*this);
+    auto weight_size = weight->data->sizes();
+    std::vector<long> kernel_size(weight_size.begin() + 2, weight_size.end());
+    auto input_size = ggI->data->sizes();
+    std::vector<long> input_shape(input_size.begin() + 2, input_size.end());
+    for(size_t i=0; i<gw_conv_params.padding.size(); ++i) {
+      // Formula for conv output size before the floor operation
+      auto out_size = float(input_shape[i] + 2 * gw_conv_params.padding[i] +
+                      gw_conv_params.dilation[i] * (kernel_size[i] - 1) - 1) /
+                      gw_conv_params.stride[i] + 1;
+      if (floorf(out_size) != out_size) {
+        // TODO: narrow ggI here to ignore these elements?
+        throw std::runtime_error("Some input elements have been lost during ConvForward"
+        " (see documentation for the Conv layer) so ConvBackwardBackward cannot be used."
+        " Resize the input so that no element is lost to be able to use ConvBackwardBackward.");
```

Comment:
In this case, we should narrow ggI and thus a Narrow function should be added.
I was wondering if there is an equivalent version of `mark_shared_storage()` for cpp functions?